src/apps/persistence/jobs.clj

```diff
@@ -645,23 +650,73 @@
           (where {:job_id job-id})
           (order :step_number)))
 
-(defn- child-job-subselect
+(defn- related-job-ids-query
+  "Returns a query that can be used to obtain the ID and parent ID of every job in the database whose ID or parent ID is
+  in the given set of job IDs. This helps to force the representative job steps query to narrow its result set down
+  early in the query execution."
   [job-ids]
-  (subselect :jobs
-             (fields :id)
-             (where {:parent_id [in job-ids]})
-             (limit 1)))
+  (let [job-ids (db/sql-array "uuid" job-ids)]
+    (-> (h/select :id :parent_id)
+        (h/from :jobs)
+        (h/where [:or
+                  [:= :id (hsql/call :any job-ids)]
+                  [:= :parent_id (hsql/call :any job-ids)]]))))
+
+(defn- batch-parent-job-ids-query
+  "Returns a query that extracts the IDs of batch parent jobs from the related_job_ids query, which should be available
+  as a common table expression (CTE). This exists just to make the representative_job_ids query a little cleaner."
+  [related-job-ids-alias]
+  (-> (h/select :id)
+      (h/from [related-job-ids-alias :p])
+      (h/where [:exists (-> (h/select :*)
+                            (h/from [related-job-ids-alias :c])
+                            (h/where [:= :p.id :c.parent_id]))])))
+
+(defn- representative-job-ids-query
+  "Returns a query that obtains the ID, parent ID, and representative job ID of every job in the related_job_ids query,
```

Comment:
Should this doc refer to the `related-job-ids-alias` param instead?
```suggestion
  "Returns a query that obtains the ID, parent ID, and representative job ID of every job in the related-job-ids-alias query,
```

Reply:
Yep, it should! Thanks for catching that.